### PR TITLE
The Rustdoc book

### PR DESCRIPTION
**Pull request by steveklabnik**
A work-in-progress start for docs for rustdoc.

This doesn't actually generate the docs yet; I wanted to open this PR to get feedback on this approach, the chapters headings themselves, and to see if anyone wanted to help fill in the ones that aren't done yet.

Start of #42322.

/cc @rust-lang/dev-tools @rust-lang/docs

### CODE REVIEW

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+```bash
+$ rustdoc -V
+$ rustdoc --version
+```
+
+This will show `rustdoc`'s version, which will look something
+like this:
+
+```text
+rustdoc 1.y.0 (hash date)
```
Comment (killercup) [inline src/doc/rustdoc/src/command-line-arguments.md]: `hash date`? Never heard of her. (I wouldn't try to make the output generic but give concrete, current outputs. People can easily see the pattern themselves.)

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+$ rustdoc -v src/lib.rs
+$ rustdoc --verbose src/lib.rs
+```
+
+This enables "verbose mode", which means that more information will be written
+to standard out. What is written depends on the other flags you've passed in.
+For example, with `--version`:
+
+```text
+$ rustdoc --version -v
```
Comment (killercup) [inline src/doc/rustdoc/src/command-line-arguments.md]: That looks weird. Either `-v -V` or `--verbose --version`

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 3]
```diff
@@ -0,0 +1,358 @@
+# Command-line arguments
+
+Here's the list of arguments you can pass to `rustdoc`:
```
Comment (killercup) [inline src/doc/rustdoc/src/command-line-arguments.md]: Maybe add a warning that these **do not** work with `cargo doc`

File: src/doc/rustdoc/src/what-is-rustdoc.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+## Using rustdoc with Cargo
+
+Cargo also has integration with `rustdoc` to make it easier to generate
+docs. Instead of the `rustdoc` command, we could have done this:
+
+```bash
+$ cargo doc
+```
+
+Internally, this calls out to `rustdoc` like this:
```
Comment (killercup) [inline src/doc/rustdoc/src/what-is-rustdoc.md]: Internally*

\* you can see this with `cargo doc --verbose`

**Review by killercup — COMMENTED** [review]
Good start. Left a few comments.

FTR, I'd really like the passes chapter to contain the needed args to render all items, even private ones. (I'm pretty sure `cargo doc -- --no-defaults --passes collapse-docs --passes unindent-comments` worked once)

**Comment by steveklabnik** [comment]
> FTR, I'd really like the passes chapter to contain the needed args to render all items, even private ones. (I'm pretty sure cargo doc -- --no-defaults --passes collapse-docs --passes unindent-comments worked once)

Yup, agreed

**Review by frewsxcv — APPROVED** [review]

File: src/doc/rustdoc/src/command-line-arguments.md [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+```bash
+$ rustdoc src/lib.rs --crate-name mycrate
+```
+
+By default, `rustodc` assumes that the name of your crate is the same name
+as the `.rs` file. `--crate-name` lets you override this assumption with
+whatever name you choose.
+
+stable(optmulti("L", "library-path", "directory to add to crate search path",
```
Comment (frewsxcv) [inline src/doc/rustdoc/src/command-line-arguments.md]: what's this doing?

Comment (frewsxcv) [inline src/doc/rustdoc/src/command-line-arguments.md]: also looks like these `stable(...)` things are used elsewhere in this file

File: src/doc/rustdoc/src/documentation-tests.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Documentation tests
+
+Coming soon!
```
Comment (frewsxcv) [inline src/doc/rustdoc/src/documentation-tests.md]: intending to ship this with the "Coming soon!" test?

File: src/doc/rustdoc/src/what-is-rustdoc.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
+# What is rustdoc?
+
+The standard Rust distribution ships with a tool called `rustdoc`. It's job is
```
Comment (frewsxcv) [inline src/doc/rustdoc/src/what-is-rustdoc.md]: `s/It's/Its`

File: src/doc/rustdoc/src/what-is-rustdoc.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+to generate documentation for Rust projects. On a fundamental level, Rustdoc
+takes as an argument either a crate root or a Markdown file, and produces HTML,
+CSS, and JavaScript.
+
+## Basic usage
+
+Let's give it a try! Let's create a new project with Cargo:
+
+```bash
+$ cargo new docs -- lib
```
Comment (frewsxcv) [inline src/doc/rustdoc/src/what-is-rustdoc.md]: is this supposed to be `--lib`? isn't this the default for `cargo new`?

File: src/doc/rustdoc/src/what-is-rustdoc.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+... and then re-run `rustdoc`:
+
+```bash
+$ rustdoc src/lib.rs --crate-name docs
+```
+
+We'll have some generated documentation. Open up `doc/docs/index.html` and
+check it out! It should show a link to the `foo` function's page, which
+is located at `doc/docs/fn.foo.html`. On that page, you'll see the "foo is
+a function" we put inside the documentation comment in our crate.
```
Comment (frewsxcv) [inline src/doc/rustdoc/src/what-is-rustdoc.md]: could also maybe `rustdoc --open` here, but not mandatory by any means

**Review by frewsxcv — CHANGES_REQUESTED** [review]
looking forward to this!

ignore my previous approval, that was an accident

File: src/doc/rustdoc/src/what-is-rustdoc.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+Now, `doc/docs/index.html` will be generated, and the page says "Crate docs."
+
+For the second issue, it's because our function `foo` is not public; `rustdoc`
+defaults to generating documentation for only public functions. If we change
+our code...
+
+```rust
+/// foo is a function
+fn foo() {}
```
Comment (matthewjasper) [inline src/doc/rustdoc/src/what-is-rustdoc.md]: Is this supposed to be `pub fn`?

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 5 above)
Comment (steveklabnik) [inline src/doc/rustdoc/src/command-line-arguments.md]: i made this page by copy/pasting the source, and must have forgotten to remove these lines 😓

File: src/doc/rustdoc/src/documentation-tests.md (hunk 6 above)
Comment (steveklabnik) [inline src/doc/rustdoc/src/documentation-tests.md]: I am happy shipping this at whatever point, this chapter should be easy to knock out though

File: src/doc/rustdoc/src/what-is-rustdoc.md (hunk 10 above)
Comment (steveklabnik) [inline src/doc/rustdoc/src/what-is-rustdoc.md]: yes 😓

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 3 above)
Comment (steveklabnik) [inline src/doc/rustdoc/src/command-line-arguments.md]: they dont?

**Comment by steveklabnik** [comment]
I've pushed up a commit that should address all rustdoc feedback so far.

**Comment by steveklabnik** [comment]
I think I would like to ship this as-is, and fill out the rest later, but if everyone else wants me to finish it before merging, that's fine as well.

File: src/doc/rustdoc/src/command-line-arguments.md (hunk 3 above)
Comment (killercup) [inline src/doc/rustdoc/src/command-line-arguments.md]: Nope:

```bash
$cargo doc --passes strip-priv-imports
error: Unknown flag: '--passes
…
```

I think it was once possible to do `cargo doc -- args…` but with latest stable:

```bash
$ cargo doc -- --passes strip-priv-imports
error: Invalid arguments.
```

Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/command-line-arguments.md]: Right, you'd have to use RUSTDOCFLAGS for this, much like RUSTFLAGS for arguments to rustc.

Comment (steveklabnik) [inline src/doc/rustdoc/src/command-line-arguments.md]: @rust-lang/cargo , is this something that's expected, or a bug? I'm kinda surprised, but I'm sure there's a good reason.

**Review by frewsxcv — APPROVED** [review]
r=me unless you have more changes

**Comment by steveklabnik** [comment]
@bors: r=frewsxcv 

nope; at least not for now.

**Pull request merged**